firestore.rules

```diff
@@ -79,71 +71,52 @@ service cloud.firestore {
-          isUser(userId)
-            || isOwner(orNull(resource.data, 'organization'))
-            || isClinician(orNull(resource.data, 'organization'))
-        )
-      );
+      function isAllowedUpdateWithinOrganization() {
+        return resource.data.type in ['patient']
+          ? isOwnerOf(resource.data.organization)
+          : isOwnerOrClinicianOf(resource.data.organization) && resource.data.type in ['clinician'];
+      }
+
+      allow read: if isAdmin()
+        || isUser(userId)
+        || isOwnerOrClinicianOf(resource.data.organization);
+
+      allow create: if isAdmin()
+        || (isUser(userId) && !('organization' in request.resource.data) && !('type' in request.resource.data));
+
+      allow update: if isAdmin()
+        || (securityRelatedFieldsDidNotChange() && (isUser(userId) || isAllowedUpdateWithinOrganization()));
 
       allow delete: if isAdmin();
     }
 
     match /users/{userId}/{collectionName}/{documentId} {
-      function isPatientWritableCollectionName() {
-        return collectionName.matches('/^[A-Za-z]+Observations$/')
-          || collectionName == 'questionnaireResponses';
+      function isOwnerOrClinicianOfSameOrganization() {
+        let user = getUser(userId);
+        return isOwnerOrClinicianOf(user.organization);
       }
 
-      function isClinicianWritableCollectionName() {
-        return collectionName == 'allergyIntolerances'
-          || collectionName == 'appointments'
-          || collectionName == 'medicationRequests'
-          || collectionName == 'medicationRecommendations'
-          || collectionName == 'symptomScores'
-          || collectionName == 'messages';
+      function isPatientWritableCollectionName() {
+        return collectionName.matches('^[A-Za-z]+Observations$')
+          || collectionName in ['questionnaireResponses']
       }
 
-      function isClinicianReadable() {
-        let user = getUser(userId);
-        return isAdmin() 
-          || isOwner(orNull(user, 'organization'))
-          || isClinician(orNull(user, 'organization'));
+      function isClinicianWritableCollectionName() {
+        return isPatientWritableCollectionName()
+          || collectionName in ['allergyIntolerances', 'appointments', 'medicationRequests'];
       }
 
-      function isClinicianWritable() {
-        let user = getUser(userId);
-        return isAdmin() || (
-          isClinicianWritableCollectionName() && (
-            isOwner(orNull(user, 'organization'))
-              || isClinician(orNull(user, 'organization'))
-          )
-        );
-      }
+      allow read: if isAdmin() 
+        || isUser(userId) 
+        || isOwnerOrClinicianOfSameOrganization();
 
-      allow read: if isAuthenticated() && (isUser(userId) || isClinicianReadable());
-      allow write: if isAuthenticated() && (
-        (isUser(userId) && isPatientWritableCollectionName())
-          || isClinicianWritable()
-      );
+      allow write: if isAdmin() 
```

Comment:
Not sure if I am missing this, but we need to ensure that the organization that is set here is only set by someone that is part of the organization (e.g. You are an owner of the Stanford Org. Same for me. As an owner of the Stanford org I shouldn't be able to add JH to your org list and make you an owner on this org).

Reply:
This is the rule for user collections (e.g. `users/{userId}/bodyWeightObservations`), not for the user itself - the user has its own match statement above with differing rules